converter.go

```diff
@@ -147,34 +148,75 @@ func valueToString(v driver.Value, tsmode snowflakeType) (*string, error) {
 		s := v1.String()
 		return &s, nil
 	case reflect.Struct:
-		if tm, ok := v.(time.Time); ok {
-			switch tsmode {
-			case dateType:
-				_, offset := tm.Zone()
-				tm = tm.Add(time.Second * time.Duration(offset))
-				s := strconv.FormatInt(tm.Unix()*1000, 10)
-				return &s, nil
-			case timeType:
-				s := fmt.Sprintf("%d",
-					(tm.Hour()*3600+tm.Minute()*60+tm.Second())*1e9+tm.Nanosecond())
-				return &s, nil
-			case timestampNtzType, timestampLtzType:
-				unixTime, _ := new(big.Int).SetString(fmt.Sprintf("%d", tm.Unix()), 10)
-				m, _ := new(big.Int).SetString(strconv.FormatInt(1e9, 10), 10)
-				unixTime.Mul(unixTime, m)
-				tmNanos, _ := new(big.Int).SetString(fmt.Sprintf("%d", tm.Nanosecond()), 10)
-				s := unixTime.Add(unixTime, tmNanos).String()
-				return &s, nil
-			case timestampTzType:
-				_, offset := tm.Zone()
-				s := fmt.Sprintf("%v %v", tm.UnixNano(), offset/60+1440)
-				return &s, nil
+		switch typedVal := v.(type) {
+		case sql.NullTime:
+			if !typedVal.Valid {
+				return nil, nil
+			}
+			return timeTypeValueToString(typedVal.Time, tsmode)
+		case time.Time:
+			return timeTypeValueToString(typedVal, tsmode)
+		case sql.NullString:
+			if !typedVal.Valid {
+				return nil, nil
+			}
+			return &typedVal.String, nil
+		case sql.NullInt64:
+			if !typedVal.Valid {
+				return nil, nil
+			}
+			s := strconv.FormatInt(typedVal.Int64, 10)
+			return &s, nil
+		case sql.NullFloat64:
+			if !typedVal.Valid {
+				return nil, nil
 			}
+			s := strconv.FormatFloat(typedVal.Float64, 'g', -1, 32)
+			return &s, nil
+		case sql.NullBool:
+			if !typedVal.Valid {
+				return nil, nil
+			}
+			s := strconv.FormatBool(typedVal.Bool)
+			return &s, nil
+		}
+		if ns, ok := v.(sql.NullBool); ok {
+			if !ns.Valid {
+				return nil, nil
+			}
+			s := strconv.FormatBool(ns.Bool)
+			return &s, nil
 		}
 	}
 	return nil, fmt.Errorf("unsupported type: %v", v1.Kind())
 }
 
+func timeTypeValueToString(tm time.Time, tsmode snowflakeType) (*string, error) {
```

Comment:
Do we have a test for this mathematics?

Reply:
We didn't had the tests for that. I have added them.